### DISCUSSION TITLE
A question about the efficiency

### DISCUSSION THREAD

**Discussion by wzm2256** [e3nn/e3nn#449] [Q&A] [answered]
Hi, thanks for the amazing library!

I have a question regarding the efficiency of the computation. It seesm most of the computation is done using jit and fx, which is of course faster than the naive implement using python for loops. My question is whether it is possible to further acclerate this? or is the usage of cuda kernel already efficient, so further improvement is not really necessary? I observed that the GPU usage is below 100% when I train the model, so I tend to believe that there are still space for improvement.

I am reading the code of SE3-transformer by Nvidia https://github.com/NVIDIA/DeepLearningExamples/tree/master/DGLPyTorch/DrugDiscovery/SE3Transformer. When all degrees have the same channel number, they stack the data tensor in the degree dimension as $(P, C, D_1+D_2)$, (P is the number of points, C is the channel, and $D_i$ is the dimension of degree-i tensor), so that the product can be done in one python step (without explicity for loop). This is useful because it can maximize the usage of cuda kernels (reshaping step add some overheads). I am wondering whether such techniques can be used in this library to accelerate the computation in some special cases.

Another possible way is to use torch.compile to accelerate the library, but it seems some issues exist https://github.com/e3nn/e3nn/discussions/431.

Any suggestion would be useful!

**Comment by mitkotak** [comment, marked as answer]
You are right on the mark there with the torch.compile reference !

Currently working on FullTensorProduct (https://github.com/e3nn/e3nn/pull/436) and Linear (https://github.com/e3nn/e3nn/pull/446). Happy to prioritize any particular features that you might need.

**Reply by wzm2256** [reply]
That looks great! I will try these in the following weeks.

**Reply by mitkotak** [reply]
Just discovered this last week but turns out you can do fullgraph=True torch.compile on TensorProduct and friends using 
```python
import e3nn
e3nn.set_optimization_defaults(jit_script_fx=False)

model = ....
model = torch.compile(model, fullgraph=True)
```

**Reply by mitkotak** [reply]
@wzm2256 Should be able to make it work with this https://github.com/e3nn/e3nn/pull/455.
